### PR TITLE
Get Scope From Headers

### PR DESCRIPTION
**Pull request by clarelimo**
## Issue Statement

When invoking a write API the client is sending the incorrect scope.

This issue is tracked at [Azure/acr](https://github.com/Azure/azure-sdk-for-net/issues/16511).

## Root Cause Analysis

In the GetScopeFromHeaders method .A split is done in the headerValue by a comma(,). This in turn excludes the push action from the scope since it has a comma before it.

## Resolution

The fix is to find the index of "scope=" in the string and get its substring.

## Tests
Manual tests have been done on a console app to verify the pull and push of a manifest Which was successful.
Automation Tests have also been done to check on more scope scenarios.

### CODE REVIEW

**Comment by jsquire** [comment]
Hi @clarelimo.  Thank you for helping to improve the Azure SDK experience and for your contribution.   I've added some tags to route this to the team best able to provide feedback.

**Comment by yugangw-msft** [comment]
@sajayantony, please take a look

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/ContainerRegistryCredentials.cs [hunk 1]
```diff
@@ -224,7 +224,7 @@ private string GetAcrAccessToken(string scope)
 
         private async Task<string> GetScope(string operation, string method, string path)
         {
-            string methodOperationKey = $"{method}>{operation}";
+            string methodOperationKey = $"{method}>{operation}";
```
Comment (avtakkar) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/ContainerRegistryCredentials.cs:227]: Remove whitespace/tabs

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/ContainerRegistryCredentials.cs [hunk 2]
```diff
@@ -280,22 +280,26 @@ private string GetScopeFromHeaders(HttpHeaders headers)
...
-                if (keyValues.Length != 2)
-                {
-                    throw new Exception($"{challengeHeader} has incorrect format, " +
-                        $"header key-value pair '{part}' does not have a value but in '{headerValue}'");
-                }
-                if (keyValues[0].ToLower().Trim() == "scope")
-                {
-                    return TrimDoubleQuotes(keyValues[1]);
-                } 
+                scope = headerValue.Substring(position);
```
Comment (avtakkar) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/ContainerRegistryCredentials.cs:292]: While this will likely work in most cases, it's not guaranteed that the server will always return a response in this format. Here's an example response from the server for the `WWW-Authenticate` header:
`Www-Authenticate: Bearer realm="https://myregistry.azurecr-test.io/oauth2/token",service="myregistry.azurecr-test.io",scope="foo:bar:baz"`
To generalize, this value's format is:
`<scheme> realm="<realm>", ..`
as shown here: https://docs.docker.com/registry/spec/api/#base
So while taking a substring, you should only read until the next `,` or if not present, the end of the string.

**Comment by sajayantony** [comment]
@avtakkar - could you recommend the test scenario so that this doesn't regress going forward.

**Comment by avtakkar** [comment]
> @avtakkar - could you recommend the test scenario so that this doesn't regress going forward.

Discussed some test cases on a call based on https://tools.ietf.org/html/rfc7235#section-4.1:
```
1. Bearer realm="https://myregistry.azurecr-test.io/oauth2/token",service="myregistry.azurecr-test.io",scope="foo:bar:baz,abc", Basic xyz=123
2. Bearer realm="https://myregistry.azurecr-test.io/oauth2/token",service="myregistry.azurecr-test.io",scope="foo:bar:baz,abc",Basic xyz=123
3. Bearer realm="https://myregistry.azurecr-test.io/oauth2/token",service="myregistry.azurecr-test.io",scope="foo:bar:baz,abc"
4. Bearer realm="https://myregistry.azurecr-test.io/oauth2/token",service="myregistry.azurecr-test.io",scope="foo:bar:baz"
5. Bearer realm="https://myregistry.azurecr-test.io/oauth2/token",service="myregistry.azurecr-test.io",scope="foo:bar:baz",abc="xyz"
```

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/ContainerRegistryCredentials.cs [hunk 3]
```diff
@@ -265,12 +266,12 @@ private string ResolveScopeLocally(string operation)
...
         /// <summary>
-        /// Parse value of scope key from the 'Www-Authenticate' challenge header. See RFC 7235 section 4.1 for more info on the 
-        /// Ex challenge header value: 
+        /// Parse value of scope key from the 'Www-Authenticate' challenge header. See RFC 7235 section 4.1 for more info on the
+        /// Ex challenge header value:
         ///  Bearer realm="https://test.azurecr.io/oauth2/token",service="test.azurecr.io",scope="repository:hello-txt:metadata_read"
         /// Return null if it is not present
         /// </summary>
-        private string GetScopeFromHeaders(HttpHeaders headers)
+        public string GetScopeFromHeaders(HttpHeaders headers)
```
Comment (avtakkar) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/ContainerRegistryCredentials.cs:274]: Add a null check since this is now `public`

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/ContainerRegistryCredentials.cs [hunk 4]
```diff
@@ -280,29 +281,31 @@ private string GetScopeFromHeaders(HttpHeaders headers)
...
             {
-                scope = headerValue.Substring(position);
+               throw new Exception($"Could not find a scope in the {headerValue}");
             }
-            else
+            else if(length > 2)
             {
-                throw new Exception($"Could not find a scope in the {headerValue}");
+                var pattern = "\".+:.+:.+\"";
+                var regex = new Regex(pattern);
```
Comment (avtakkar) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/ContainerRegistryCredentials.cs:302]: It would be better to make `pattern` a `const` and `regex` and a `private static readonly` member.

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/ContainerRegistryCredentials.cs [hunk 5]
```diff
@@ -280,29 +281,31 @@ private string GetScopeFromHeaders(HttpHeaders headers)
...
                 }
             }
 
-            string scope = "";
             int position = headerValue.IndexOf("scope=");
-            if (position > 0)
+            string scope = headerValue.Substring(position);
+            string[] keyValues = scope.Split('=');
+            int length = keyValues.Length;
+            var scopeContainedIn = keyValues[1];
```
Comment (avtakkar) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/ContainerRegistryCredentials.cs:293]: At this point, the length hasn't been checked so you could get a `IndexOutOfRangeException`

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/tests/Tests/ScopeTests.cs [hunk 6]
```diff
@@ -0,0 +1,147 @@
...
+                }
+            }
+        };
+
+        private void VerifyManifest(Manifest baseManifest, Manifest actualManifest)
+        {
+            Assert.Equal(baseManifest.GetType(), actualManifest.GetType());
+            Assert.Equal(baseManifest.SchemaVersion, actualManifest.SchemaVersion);
+
+            if (baseManifest.GetType() == typeof(V2Manifest))
```
Comment (avtakkar) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/tests/Tests/ScopeTests.cs:92]: Use pattern matching with `is`: https://docs.microsoft.com/en-us/dotnet/csharp/language-reference/keywords/is
Example: `baseManifest is V2Manifest v2Manifest`

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/ContainerRegistryCredentials.cs [hunk 7]
```diff
@@ -273,9 +275,13 @@ private string ResolveScopeLocally(string operation)
         /// </summary>
         public string GetScopeFromHeaders(HttpHeaders headers)
         {
+            if (headers == null)
+            {
+                throw new ValidationException(ValidationRules.CannotBeNull, "headers");
```
Comment (avtakkar) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/ContainerRegistryCredentials.cs:280]: `nameof(headers)`

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/ContainerRegistryCredentials.cs [hunk 8]
```diff
@@ -54,6 +54,8 @@ public enum LoginMode
         // Internal simplified client for Token Acquisition
         private ContainerRegistryRefreshToken _acrRefresh;
         private AuthToken _aadAccess;
+        private const string pattern = "\".+:.+:.+\"";  //Pattern to get the scope from headers
+        private static readonly Regex regex = new Regex(pattern);
```
Comment (avtakkar) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/src/Customizations/ContainerRegistryCredentials.cs:58]: Either document what this is, or just rename to something more descriptive. For instance, `scopeFromHeaderRegex`

File: sdk/containerregistry/Microsoft.Azure.ContainerRegistry/tests/Tests/ScopeTests.cs [hunk 9]
```diff
@@ -104,42 +104,6 @@ private void VerifyManifest(Manifest baseManifest, Manifest actualManifest)
                 Assert.Equal(baseManifestV2.Config.MediaType, actualManifestV2.Config.MediaType);
                 Assert.Equal(baseManifestV2.Config.Size, actualManifestV2.Config.Size);
             }
-            if (baseManifest.GetType() == typeof(V1Manifest))
```
Comment (avtakkar) [inline sdk/containerregistry/Microsoft.Azure.ContainerRegistry/tests/Tests/ScopeTests.cs]: Why is this removed?

Comment (clarelimo) [reply sdk/containerregistry/Microsoft.Azure.ContainerRegistry/tests/Tests/ScopeTests.cs]: I am only using V2Manifest in the tests.

**Review by yugangw-msft — APPROVED** [review]
